src/imitation/algorithms/preference_comparisons.py

```diff
@@ -919,16 +963,43 @@ def _make_data_loader(self, dataset: PreferenceDataset) -> data_th.DataLoader:
 
     def _train(self, dataset: PreferenceDataset, epoch_multiplier: float = 1.0) -> None:
         """Trains for `epoch_multiplier * self.epochs` epochs over `dataset`."""
-        dataloader = self._make_data_loader(dataset)
+        if self.regularizer:
+            val_length = int(len(dataset) * self.val_split)
+            train_length = len(dataset) - val_length
+            train_dataset, val_dataset = data_th.random_split(
+                dataset,
+                lengths=[train_length, val_length],
+                generator=th.Generator().manual_seed(self.seed),
+            )
+            dataloader = self._make_data_loader(train_dataset)
+            val_dataloader = self._make_data_loader(val_dataset)
+        else:
+            dataloader = self._make_data_loader(dataset)
+            val_dataloader = None
+
         epochs = round(self.epochs * epoch_multiplier)
 
         for _ in tqdm(range(epochs), desc="Training reward model"):
+            train_loss = 0.0
+            val_loss = 0.0
             for fragment_pairs, preferences in dataloader:
                 self.optim.zero_grad()
                 loss = self._training_inner_loop(fragment_pairs, preferences)
-                loss.backward()
+                train_loss += loss.item()
+                if self.regularizer:
+                    self.regularizer.regularize(loss)
                 self.optim.step()
 
+            if self.regularizer:
+                # TODO(juan) my type checker complains because it cannot infer the type
+                #  of val_dataloader by checking the logic. Is there a better way
+                #  to do this, or should we leave the cast?
+                val_dataloader = cast(data_th.DataLoader, val_dataloader)
+                for fragment_pairs, preferences in val_dataloader:
+                    loss = self._training_inner_loop(fragment_pairs, preferences)
+                    val_loss += loss.item()
+                self.regularizer.update_params(train_loss, val_loss)
```

Comment:
To check semantics: `train_loss` and `val_loss` are summed across entire dataset? 

Guess that's fine, scale doesn't really matter given you're just doing relative comparison between train and val loss.

Reply:
Yes, that's what I'm doing, I was not sure which approach to follow. If you have a preference for the other, let me know. I might have to change this anyways given the other issue/PR on batch size.